Propose a feature to troubleshoot running pods

verb opened the pull request:
This feature allows troubleshooting of running pods by running a new "Debug Container" in the pod namespaces.

This proposal was originally opened and reviewed in kubernetes/kubernetes#35584.

This proposal needs LGTM by the following SIGs:
- [ ] SIG Node
- [ ] SIG CLI
- [ ] SIG Auth
- [x] API Reviewer

Work in Progress:
- [x] Prototype `kubectl attach` for debug containers
- [x] Talk to sig-api-machinery about `/debug` subresource semantics

verb commented:
/assign @dashpole

dashpole commented:
/lgtm

derekwaynecarr commented:
I think this proposal needs wider review since it comes with API changes.

derekwaynecarr commented:
Related to feature https://github.com/kubernetes/features/issues/277

derekwaynecarr commented on contributors/design-proposals/troubleshoot-running-pods.md:
````diff
@@ -0,0 +1,570 @@
+```
+type PodStatus struct {
+        ...
+        DebugContainerStatuses []ContainerStatus
````
since this feature is alpha, doesn't the field name need to reflect that?

verb replied:
That would make sense, I just didn't know that was how to do it. I know there's been a lot of discussion in #30819, but I haven't followed it. It doesn't look like there's consensus, but I can try to cherry-pick something that might work.

gyliu513 replied:
I think that we do not need to update the filed name but how about using `annotations` like `init container`? Such as adding an annotation as `pod.alpha.kubernetes.io/debug-containers`? You can get more detail from https://github.com/kubernetes/community/blob/master/contributors/design-proposals/container-init.md

liggitt replied:
We should definitely not store something as important as debug container info in an annotation. Those are also compatibility nightmares when they transition to supported fields. 

@kubernetes/sig-api-machinery-misc for guidance on alpha field names and gating

gyliu513 replied:
@liggitt can you please show more detail for this? I found that the `init container` was using this logic to be from alpha->beta->graduate, why cannot the `debug container`?

liggitt replied:
It did, and it has been a very painful transition, with very poor user experience (c.f. https://github.com/kubernetes/kubernetes/issues/45627#issuecomment-301863792)

gyliu513 replied:
Got it, thanks @liggitt

derekwaynecarr commented:
I think a section that covers the security concerns associated with this feature would be appreciated for future readers.

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+
+`startContainer()` will be updated to write a new label
+`io.kubernetes.container.type` to the runtime. Existing containers will be
+started with a type of `REGULAR` or `INIT`. When added in a subsequent step,
```
does this mean a 1.7 kubelet managing containers without this label (that were started by a 1.6 kubelet) will be confused about whether they are debug containers?

verb replied:
No, it's backwards compatible (and there's a test). The `REGULAR` and `INIT` labels aren't used by anything initially and the kubelet behavior only differs when `Type == "DEBUG"`. (Type will be an empty string for a container that existed prior to the feature being enabled.)

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+
+### Additional Constraints
+
+1.  Non-interactive workloads are explicitly supported. There are no plans to
```
If the container run and attach steps are distinct, how is stdout/stderr coordinated so that the attach request obtains the first byte written to each? Is the first attach special? How do subsequent or additional attaches behave w.r.t. previous output from the debug container's entrypoint?

liggitt replied:
if @ncdc has learned anything, it is that if an output buffer race can happen, it will, and that if it might be a problem, it will.

ncdc replied:
@liggitt the typical process for starting a docker container is (outside of kubernetes):

1. create container
2. attach to container
3. start container

If you must start the container prior to attaching (which tends to be the case for things likes `kubectl run`), then your only option to make sure you see all prior output is to specify `logs=true` when attaching. This has downsides: last time I checked, you can't limit the output to e.g. the last 100 lines, and if you have a TTY, I'm not sure what happens in that case. Also note, this isn't currently available in the version of the docker api vendored in to kubernetes.

ncdc replied:
I just wrote in another comment about this, but I think given the way the kubelet sync loop works, it will be very difficult to achieve create/attach/start without blocking one of the kubelet workers. We will need to discuss with sig-node if we want to pursue this.

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+1.  Non-interactive workloads are explicitly supported. There are no plans to
+    supported detached workloads, but doing so would be trivial with an
+    `attach=false` flag.
+1.  There are no guaranteed resources for ad-hoc troubleshooting. If
```
this seems like it would make debugging a pod that was memory constrained pretty difficult.

liggitt commented:
how do debug containers interact with graceful termination of pods?

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+    policy.
+*   Explicit reattaching isn't implemented. Instead a `kubectl debug` invocation
+    will implicitly reattach if there is an existing, running container with the
+    same name. In this case container configuration will be ignored.
```
I agree 100%

liggitt replied:
so if this moves from alpha to beta to stable, how do you maintain skew compatibility with older kubectl clients' debug implementations without propagating the implicit throw-away-info-and-reattach behavior into the stable version of the feature?

ncdc replied:
I think if we just expand `kubectl attach` to support debug containers, that would solve the problem.

verb replied:
either way this has to be resolved prior to moving out of alpha

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+after populating `kubecontainer.ContainerStatus`, the label value will be stored
+in a the new field `ContainerStatus.Type` so it can be used by `SyncPod()`.
+
+The kubelet gains a `RunDebugContainer()` method which accepts a `v1.Container`
```
Yes, the pod spec is needed to generate the podsandbox configuration. It wasn't my intention to imply that a v1.Container was all that is required. Here's a staged copy of the change:

https://github.com/verb/kubernetes/pull/3/files#diff-44f60dd6d99cb695e5f333647ebd0703R753

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+*   Resources
+*   LivenessProbe
+*   ReadinessProbe
+*   Lifecycle
```
There's nothing inherently incompatible about volume mounts in a Debug Container, though there might be edge cases.

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+
+### Killing Debug Containers
+
+Debug containers will not be killed automatically until the pod (specifically,
```
That didn't occur to me as something we would want, though it does seem like a reasonable compromise. I'll look into it.

verb commented:
@liggitt Debug containers receive the same signals as other containers in the pod for lifecycle events. They differ only in that they aren't deleted when syncing pod spec while the pod is alive.

verb commented:
/assign @pwittrock 
/assign @liggitt

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+
+1.  `kubectl` constructs a `v1.Container` based on command line flags and
+    `POST`s it to `/api/v1/namespaces/$NS/pods/$POD/debug`.
+1.  The API server performs admission control and proxies the connection to the
```
> I would hope the plugins will deny resources they don't recognize.

they do not, they select the resources/subresources they guard.

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+    status
+*   Exited Debug Containers will be garbage collected as regular containers and
+    may disappear from the list of Debug Container Statuses.
+*   There's no specific integration with admission controller and pod security
```
API resources are registered in https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/rest/storage_core.go#L187. Omitting an item from that map means it is not surfaced via the API.

liggitt commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+    sandbox.
+1.  `convertStatusToAPIStatus()` must sort Debug Containers status into
+    `DebugContainerStatuses` similar to as it does for `InitContainerStatuses`
+1.  The kubelet must preserve information on debug containers for reporting.
```
the entire v1.Container spec is placed into a label?

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
````diff
@@ -0,0 +1,570 @@
+```
+
+It would be reasonable for Kubernetes to provide a default container name and
+image1, making the minimal possible debug command:
````
image

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
````diff
@@ -0,0 +1,570 @@
+image1, making the minimal possible debug command:
+
+```
+kubectl debug -it target-pod
````
Do you think you'd ever want to run `kubectl debug` and _not_ attach stdin and use a tty?

verb replied:
Updated the document to specify that -i & -t will be enabled by default.

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+This creates an interactive shell in a pod which can examine and signal all
+processes in the pod. It has access to the same network and IPC as processes in
+the pod. It can access the filesystem of other processes by `/proc/$PID/root`,
+and enter aribitrary namespaces of another container via `nsenter` when
```
Kubernetes master with a new enough docker version uses a shared pid namespace by default now.

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+replace a Debug Container that has exited by re-using a Debug Container name. It
+is an error to attempt to replace a Debug Container that is still running.
+
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
```
`kubectl attach` supports attaching to both init and normal containers. Would you want to expand it to support debug containers too? That would require the least amount of coding.

ncdc replied:
The only true way to ensure you don't miss any output is create-container, attach-container, start-container, in that order. That's how `docker run` works. For something like `kubectl run` and probably `kubectl debug` too, we can't really do that, because of the way the kubelet sync loop works (kubectl waits until it sees the pod is Running before attaching). Well, I guess we _could_ potentially do that, but it would require pausing a sync loop iteration until the remote client (kubectl) attaches, which isn't ideal.

verb replied:
Added this info to the doc, thanks!

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+### Killing Debug Containers
+
+Debug containers will not be killed automatically until the pod (specifically,
+the pod sandbox) is destroyed. Unlike `kubectl exec`, Debug Containers will not
```
There is roundabout support in newer docker versions for killing exec sessions. It now records the pid of the process that was exec'd, and we could use that information to do a kill.

derekwaynecarr replied:
@ncdc it requires docker 1.13+ i think, correct?

ncdc replied:
@derekwaynecarr yes, plus we'd need to bump the docker APIs in kube (no small feat, since kube currently uses the now-deprecated engine-api repo).

ncdc commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+the pod sandbox) is destroyed. Unlike `kubectl exec`, Debug Containers will not
+receive an EOF if their connection is interrupted. Instead, Debug Containers
+must be reattached to exit a running process. This could be tricky if the
+process does not allocate a TTY, in this case a second Debug Container could be
```
Same comment as above re why would you not allocate a tty?

ncdc replied:
I would view `kubectl debug` as an interactive utility for debugging a running pod, whereas I see `kubectl exec` as a tool in my toolbox that might or might not require user interaction. Although the more I think about it, typically when I do use `kubectl exec` it's to get a shell, in which case I almost always do `-it`. There's a recent issue proposing that exec defaults these to true: #46300.

verb replied:
@ncdc @liggitt since `kubectl debug` is mainly a tool for getting a shell with a TTY, how much do we care about output stream coordination when `kubectl` will already say "press enter to get a prompt"? I think it's reasonable to defer to a future general solution rather than trying to solve this for debug.

ncdc replied:
No objections to deferring

ericchiang commented on contributors/design-proposals/troubleshoot-running-pods.md:
````diff
@@ -0,0 +1,570 @@
+kubectl debug -it target-pod
+```
+
+This creates an interactive shell in a pod which can examine and signal all
````
What does "in a pod" mean? Each container has a different rootfs and aren't individual containers in a pods in unique PID namespaces?

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+
+The process for creating a Debug Container is:
+
+1.  `kubectl` constructs a `v1.Container` based on command line flags and
```
Added description of a new object `PodDebugContainer`

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+    Debug Containers: it creates a new runtime label (e.g. a docker label) that
+    identifies this container as a Debug Container.
+1.  The kubelet performs an attach operation and upgrades the client connection
+    to streaming.
```
Correct, this shows up in pod status shortly after the container is created (assuming api server is reachable, etc). I've updated the doc to specify this.

Re: attach, I've updated the doc to clarify that the existing attach handler and protocol semantics are used.

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,570 @@
+##### Step 3: kubelet API changes
+
+The kubelet gains a new streaming endpoint `/debugPod` and `server.getDebug()`
+which accepts a `v1.Container` via `POST` and calls `RunDebugContainer()` if
```
Done.

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,649 @@
+of Debug Containers is reported via a new field in `v1.PodStatus`, described in
+a subsequent section.
+
+#### Alternative: Extending `/exec`
```
@dchen1107 @lavalamp @smarterclayton @pwittrock 

For pod troubleshooting we need to choose between the object-based approach described above (suggested by @smarterclayton ) and the exec-based approach described below (suggested by @lavalamp and resembling the "image exec" approach of the original proposal, interestingly).

mikedanese commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,666 @@
+   13 ?        Ss     0:00 bash
+   26 ?        Ss+    0:00 /neato
+  107 ?        R+     0:00 ps x
+root@debug-image:~# cat /proc/26/root/etc/resolv.conf
```
nit: I don't think this works with docker.

verb commented:
@mikedanese it has worked on the docker installs I've tried. Could you
provide details of the install where it doesn't work? I'd like to
investigate.

On Fri, Jun 23, 2017, 13:54 Mike Danese <notifications@github.com> wrote:

> *@mikedanese* commented on this pull request.
> ------------------------------
>
> In contributors/design-proposals/troubleshoot-running-pods.md
> <https://github.com/kubernetes/community/pull/649#discussion_r123730346>:
>
> > +1.  Inspect the filesystem of target container
> +1.  Execute debugging utilities not included in the container image
> +1.  Initiate network requests from the pod network namespace
> +
> +This is achieved by running a new "debug" container in the pod namespaces. His
> +troubleshooting session might resemble:
> +
> +```
> +% kubectl debug -it -m debian neato-5thn0 -- bash
> +root@debug-image:~# ps x
> +  PID TTY      STAT   TIME COMMAND
> +    1 ?        Ss     0:00 /pause
> +   13 ?        Ss     0:00 bash
> +   26 ?        Ss+    0:00 /neato
> +  107 ?        R+     0:00 ps x
> +root@debug-image:~# cat /proc/26/root/etc/resolv.conf
>
> nit: I don't think this works with docker.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/649#pullrequestreview-45964008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcAMM93SkF-ouLe290SL-6axXXrGdNrks5sG6dvgaJpZM4NinaN>
> .
>

verb commented:
Feedback from API review with @lavalamp:

1. Don't use `debug` anywhere in the API. API names should reflect behavior rather than intended use.
2. This is not a good use case for a new imperative call that's passed-through to the kubelet. From an API perspective it would be better to have these debug containers reflected in the spec, which could be updated with a PATCH.

The consensus of sig-node is that the general container spec should be immutable, so we're at an impasse.

We could introduce a new field in the spec, something like `RunOnceContainers []Container`, and make that the only field allowed to be patched, but anything declared in the spec is easily mistaken for a regular container. From the perspective of the kubelet, this is not a general purpose container. It is an ephemeral container with no lifecycle or resource guarantees.

If we were to add to the additional complexity to the kubelet to make these general purpose, then we just have in the spec a container list that you're allowed to update and the one you're not allowed to update.

I'm happy to not use the word "debug" in the API, but I'm trying to build a better `exec` and I think the imperative call is justified. We need a compromise @derekwaynecarr @dchen1107 @vishh @smarterclayton

lavalamp commented:
Allowing mutations to containers marked `ephemeral: true` sounds pretty
good to me, actually. Documentation can clarify that ephemeral containers
have very few guarantees, etc...

On Fri, Jul 7, 2017 at 6:23 AM, Lee Verberne <notifications@github.com>
wrote:

> Feedback from API review with @lavalamp <https://github.com/lavalamp>:
>
>    1. Don't use debug anywhere in the API. API names should reflect
>    behavior rather than intended use.
>    2. This is not a good use case for a new imperative call that's
>    passed-through to the kubelet. From an API perspective it would be better
>    to have these debug containers reflected in the spec, which could be
>    updated with a PATCH.
>
> The consensus of sig-node is that the general container spec should be
> immutable, so we're at an impasse.
>
> We could introduce a new field in the spec, something like RunOnceContainers
> []Container, and make that the only field allowed to be patched, but
> anything declared in the spec is easily mistaken for a regular container.
> From the perspective of the kubelet, this is not a general purpose
> container. It is an ephemeral container with no lifecycle or resource
> guarantees.
>
> If we were to add to the additional complexity to the kubelet to make
> these general purpose, then we just have in the spec a container list that
> you're allowed to update and the one you're not allowed to update.
>
> I'm happy to not use the word "debug" in the API, but I'm trying to build
> a better exec and I think the imperative call is justified. We need a
> compromise @derekwaynecarr <https://github.com/derekwaynecarr> @dchen1107
> <https://github.com/dchen1107> @vishh <https://github.com/vishh>
> @smarterclayton <https://github.com/smarterclayton>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/649#issuecomment-313679391>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglnJdPT1bKuGuaop28Dit8uC5APa4ks5sLjFSgaJpZM4NinaN>
> .
>

derekwaynecarr commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,666 @@
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to shared pod namespaces and the file systems of individual
+    containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
```
fetch troubleshooting utilities at debug time rather than at the time of pod creation because...
- want to avoid image pulling?
- concerns about pod startup latency?

more detail here would be helpful.

sjenning replied:
Yes, if this requirement is dropped, this becomes a lot simpler.  I know that istio is looking to enable sidecarring through an admission controller server-side https://github.com/istio/kube-inject (no code yet).  Currently they have a tool, `istio kube-inject`, that injects the istio sidecar into the pod spec client-side.

verb replied:
@sjenning that's a clever solution that hadn't occurred to me, but I think it's solving a different problem. That sounds like a solution to a development or platform integration problem. Besides, let's not drop requirements to just make implementation easier. Let's solve the harder problem and delight the user.

Debug containers is an attempt to solve an operations and security problem. The operations problem is that I want to troubleshoot a running minimal container image (built from SCRATCH) without restarting it or having had to predict in advance which tools I would need.

Adding the container at pod creation has the following downsides:

- One of the motivations for minimal container images is to avoid deploying a full OS stack with every application instance to reduce attack surface.
- Specifying the debugging image at debug time allows freedom of choice between arbitrary debugging solutions. For example, I could choice between a "take snapshot and restart" container image or one that gives me an interactive shell.
- Choice of debugging image enables more useful cluster policy. For example, the "take snapshot and restart" image doesn't generate an audit alert while an interactive shell requires a human to review.
- Due to constraints of the runtime, a debugging container is only useful when it's running. Running an idle shell in hundreds of thousands of pods is inefficient and, frankly, feels like a hack.
- I want the latest debug tools rather than the old ones from when this pod was created.

These things are already mentioned in the proposal, but I understand that the proposal is too large for people to read because I've integrated hundreds (160 at last count, but that was before this PR was opened) of comments and suggestion.

Clearly I could be better at organizing information for this audience, and I would certainly appreciate any suggestions to that effect.

pwittrock commented:
/unassign

derekwaynecarr commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,666 @@
+        metav1.TypeMeta
+
+        // Template defines the pods that will be created from this pod template
+        Container v1.Container
```
per our discussion, I do not support this being a v1.Container as that brings in resource requirements concerns which are not required for this use case (unless I am mistaken).  For example, I do not imagine a debug container would need access to its own GPU device.

verb replied:
@derekwaynecarr I had planned on handling this with validation that the resources section is empty so that people can reuse the familiar `v1.Container`, but it's certainly not necessary and I'm happy to use a simpler object.

derekwaynecarr replied:
@verb - it may be worth to just have a small section explaining it.  for example, i would not imagine that a debug container would change the QoS class of a pod.

@sjenning @vishh -- if this feature was enabled, and i had static cpu pinning enabled on the kubelet, what cpu should run this pod if it was a G workload?

ConnorDoyle replied:
**Re: cpu manager**

If:
* Static cpu manager policy is enabled
* Pod QoS classification is Guaranteed
* Container requests an integer number _N_ of CPUs

Then the container will run in its own CPU set spanning _N_ logical CPUs.

Otherwise, it will run in the shared pool together with other Burstable, Best-Effort, and non-integral-cpu Guaranteed containers.

sjenning replied:
@derekwaynecarr yes, to add to @ConnorDoyle comment, the decision to pin is a container-level decision, not pod-level.  So an added container will just be subjected to the same rules as the containers in the pod spec.

derekwaynecarr commented:
this feature also has potential challenges with autoscaling, cc @DirectXMan12

verb commented:
Updated proposal with results of API review

derekwaynecarr commented:
This still requires shared pid, should we rally on this proposal to identify a plan to address concerns that prevented that being the default in 1.8?

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+order to make progress, we will start with the smallest possible API change,
+extending `/exec` to support Debug Containers, and iterate.
+
+From the perspective of the user, there's a new command, `kubectl debug`, that
```
I'd like to ensure any work on a debug command is compatible with future copy pod behavior (which is useful in many cases where this particular command is not).

verb replied:
Agreed, this is important to me, too. Last I heard the desire for copy pod was to implement it entirely client side in `kubectl run`. It was important to SIG CLI to keep the concepts separate. As of right now I know of no conflict.

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+`DebugName`.
+
+Debug Containers cannot be removed via the API and instead the process must
+terminate. While not ideal, this parallels existing behavior of `kubectl exec`.
```
kubectl exec in practice ends up closing commands early because those commands fail to write to stdout / stderr and then exit.

verb replied:
Sorry, I'm not following. You're saying a common occurrence is that commands are getting errors when writing to stdout and so they immediately exit? Is it because users are exec'ing without Stdout=False?

We were hoping with Debug Containers to remove some of the confusion around stdin/stdout/stderr/TTY by specifying them as interactive containers that always have input/output streams and TTYs. Is this in line with your concerns?

I'll have to give some thought to the interaction with the other params in PodExecOptions, though.

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+Debug Containers cannot be removed via the API and instead the process must
+terminate. While not ideal, this parallels existing behavior of `kubectl exec`.
+To kill a Debug Container one would `attach` and exit the process interactively
+or create a new Debug Container to send a signal with `kill(1)` to the original
```
That's pretty annoying / ugly.  If a user needs to guarantee the process exits, then debug containers become a significant liability.

verb replied:
Agreed, but I've been stuck for almost a year. Lots of people have insightful, valid and conflicting opinions about what the API should be. I know it's not ideal, but can we start with this and iterate?

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+
+One can reattach to a Debug Container using `kubectl attach`. When supported by
+a runtime, multiple clients can attach to a single debug container and share the
+terminal. This is supported by Docker.
```
What about other container runtimes?

verb replied:
I haven't tested them. I think this feature is useful enough to leave it open (at least for the alpha) and address problems in the runtimes. The runtimes don't have to support this, but I think it's reasonable to expect them to handle it gracefully since the situation is not unique to debug containers.

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+Debug containers will not be killed automatically until the pod (specifically,
+the pod sandbox) is destroyed. Debug Containers will stop when their command
+exits, such as exiting a shell. Unlike `kubectl exec`, processes in Debug
+Containers will not receive an EOF if their connection is interrupted.
```
Why not?

verb replied:
By design. We want to support reattaching after an interrupted connection.

smarterclayton commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+
+**Access to all pod namespaces and the file systems of individual containers.**
+This solution runs a container in the shared pod namespaces. It relies on the
+behavior of `/proc/<pid>/root` (and therefore shared PID namespace) to provide
```
I'm not convinced that most people are going to move to a shared pid namespace rapidly (because it's a significant change to how a multi container pod works).  Can you articulate why the solution requires a shared pid namespace for the pod and not just having the CRI share the pid namespace of the target container?

verb replied:
We'll no longer require the entire pod to share the PID namespace. It will work as you describe.

When this part was written pod PID namespace sharing had already been implemented for all pods and we weren't expecting objections. I'll update it.

verb commented:
@derekwaynecarr It requires PID namespace sharing, but not necessarily with the entire pod. When the pod isn't sharing a single PID namespace the Debug Container will instead join the PID namespace of the target container and it will still be useful. I think it's safe for these two to proceed in parallel.

When namespace sharing is completely unavailable (e.g. docker < 1.13.1) then I think we should fail to create the Debug Container with an obvious error message (not supported by this runtime, etc) rather than create a crippled Debug Container in order to avoid confusing the user and/or making a bad impression.

verb commented on contributors/design-proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+particular, they should enforce the same container image policy on the `Image`
+parameter as is enforced for regular containers. During the alpha phase we will
+additionally support a container image whitelist as a kubelet flag to allow
+cluster administrators to easily constraint debug container images.
```
@derekwaynecarr I added a image whitelist as a flag to address your concerns about container image provenance, but I'm not sure I completely understand them. Could you tell me if this addresses your concerns sufficiently for the alpha? One will be able to, for example, allow only debian and alpine debug container images with something like `--debug-image-whitelist=debian,alpine`

verb commented:
Agreement from sig node yesterday is this can be submitted after @lavalamp signs off on API. I believe I've addressed all feedback. If I've lost track of anything please let me know.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,739 @@
+        ...
+        // DebugName is the name of the Debug Container. Its presence will cause
+        // exec to create a Debug Container rather than performing a runtime exec.
+        DebugName string `json:"debugName,omitempty" ...`
```
I vote for making a sub section:

```go 
type PodExecOptions struct {
  ...
  EphemeralContainer *PodExecEphemeralContainerSpec
}

type PodExecEphemeralContainerSpec struct {
  Name string
  Image string
}
```

verb replied:
SGTM, though in [my naive prototype](https://github.com/verb/kubernetes/commit/c6a72efad4c72b2f7815c07d295b7ac9141199f5#diff-c73f80ad83608f18657d22a06950d929R475) the sub section wasn't populated from the HTTP params. 

I'm not very familiar with the api machinery so I probably just missed something. I see where [queryparams flattens the struct based on JSON names](https://github.com/kubernetes/apimachinery/blob/master/pkg/conversion/queryparams/convert.go#L129). Do I need to write a custom converter somewhere to get it back from params to an object? Could you point me in the right direction?

liggitt replied:
yeah, nested structs parsed from query params won't work cleanly (https://github.com/kubernetes/kubernetes/issues/21476)

lavalamp replied:
Earlier in the text it said this was not using query params--I thought it was using a message sent after the SPDY channel was opened.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,749 @@
+We will extend `v1.Pod`'s `/exec` subresource to support "executing" container
+images. The current `/exec` endpoint must implement `GET` to support streaming
+for all clients. We don't want to encode a (potentially large) `v1.Container` as
+an HTTP parameter, so we must extend `v1.PodExecOptions` with the specific
```
Oh, I see. I misunderstood this line.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,749 @@
+type PodExecOptions struct {
+        ...
+        // Run Command in an ephemeral container which shares some namespaces with Container.
+        EphemeralContainer PodExecEphemeralContainerSpec
```
This needs to be a pointer. Uh, but I will suggest something else since sub structs are indeed currently broken.

lavalamp replied:
```go
// PodExecOptions is the query options to a Pod's remote exec call
type PodExecOptions struct {
  ...
  // EphemeralContainerName is the name of an ephemeral container in which the
  // command ought to be run. Either both EphemeralContainerName and
  // EphemeralContainerImage fields must be set, or neither.
  EphemeralContainerName *string `json:"ephemeralContainerName,omitempty" ...`

  // EphemeralContainerImage is the image of an ephemeral container in which the command
  // ought to be run. Either both EphemeralContainerName and EphemeralContainerImage
  // fields must be set, or neither.
  EphemeralContainerImage *string `json:"ephemeralContainerImage,omitempty" ...`
}
```

verb replied:
Renamed as suggested

lavalamp commented:
/lgtm

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
````diff
@@ -0,0 +1,743 @@
+```
+type PodStatus struct {
+        ...
+        DebugStatuses []DebugStatus
````
should this mirror the exec option names? ephemeral statuses, etc?

smarterclayton replied:
We should use consistent naming

lavalamp replied:
Argh, yes, I was in a rush and didn't see this section. @verb can you modify to

```go
type PodStatus struct {
  
  EphemeralContainers []EphemeralContainerStatus
}
```

lavalamp replied:
Hm. Actually were you trying to represent all exec actions with this?

lavalamp replied:
(note I edited my comment above. I can't see anything you need that isn't already in v1.ContainerStatus.)

verb replied:
@lavalamp I'd like to at least have command and args, which aren't part of ContainerStatus.

smarterclayton commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+*   Exited Debug Containers will be garbage collected as regular containers and
+    may disappear from the list of Debug Container Statuses.
+*   Security Context for the Debug Container is not configurable. It will always
+    be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.
```
So only a cluster admin should ever use debug containers?

smarterclayton replied:
Also, pods don't have security context set in many cases.  Exec implicitly escalating to root is bad.

verb replied:
This should be configurable at debug time, but that depends on the API. Some of my proposed API changes addressed this, but SIG Node and the API reviewers deadlocked on which was best. Our compromise was to proceed to alpha with the minimum possible API change, which doesn't include a configurable security context.

Created https://github.com/kubernetes/kubernetes/issues/53188 to track this.

/cc @thockin

liggitt replied:
> Our compromise was to proceed to alpha with the minimum possible API change, which doesn't include a configurable security context.

you have a pointer to where that discussion happened? no one from the auth/psp side was involved afaik

verb replied:
I don't understand, you've been involved since the first draft of the proposal 10 months ago, and your input has always been most welcome. It's not too late, what would you like to see changed?

liggitt replied:
> It's not too late, what would you like to see changed?

there has to be control over the security aspects (uid/gid, capabilities, pid/network/ipc namespace, etc) of containers that are launched on the nodes. Omitting that control ("Security Context for the Debug Container is not configurable"), and defaulting to privileged levels ("It will always be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.") is a particular problem.

verb replied:
Agreed, and I created an issue to track adding a configurable security context. Until then this feature is an alpha feature requiring cluster admins to enable it.

I don't think it's a good idea to encode an entire securityContext as an HTTP query parameter. Are there any fields you can't live without during the alpha? What about a toggle for SYS_ADMIN?

liggitt replied:
> I created an issue to track adding a configurable security context. Until then this feature is an alpha feature requiring cluster admins to enable it.

When there are known gaps required for exiting alpha status that have yet-undetermined implications on the API for this feature, it doesn't seem wise to start making changes to stable APIs in master.

> I don't think it's a good idea to encode an entire securityContext as an HTTP query parameter. Are there any fields you can't live without during the alpha? What about a toggle for SYS_ADMIN?

I agree that encoding is ugly, and I don't a great idea there, but I think it's a mistake to not deal with this in terms of a standard `Container` definition:
* many parts of the security context has utility in some debug scenario:
  * running many debug tools require elevated capabilities
  * other scenarios would want to run as a specific uid/gid (plenty of overlap with https://github.com/kubernetes/kubernetes/issues/30656)
* admission plugins that make decisions on pods do so in terms of PodSpec and `Container` structs. What is the benefit of coming up with a different structure to describe the same thing?

thockin replied:
Jordan,

Is your concern mainly that the structure is not a v1.Container ?  Or is the concern around the (ab)use of /exec?  I am not super keen on the /exec bits, but I want this to unstick and forward progress feels like a better tradeoff than another LITERAL year of wheel-spinning.

If we work through the changes required to plumb security down, even though not fully implementing for alpha, does that satisfy?  If not, can you explain exactly what you'd like to see for forward motion?

liggitt replied:
> Is your concern mainly that the structure is not a v1.Container ? Or is the concern around the (ab)use of /exec? 

Mostly the former (the latter is ugly, but not fatal). I have three main concerns:

1. This proposes to build the container spec implicitly in the kubelet, given an image. That means admission doesn't have the information needed to allow/deny/modify that container spec, and if the kubelet changes how it defines the container spec release-to-release, admission can't even know what the kubelet will do.
2. There's not a clear path forward to support more aspects of the container spec (control over pull policy, uid/gid, capabilities, env, etc, all seem likely and reasonable)
3. The proposed defaults don't match the defaults for Container specs (specifically around security settings / capabilities), which would make transitioning to dealing with this as a `v1.Container` difficult

If we don't deal with this as a `v1.Container` consistently through the stack (api/admission/kubelet), we're making every layer have to understand two ways containers can be expressed, which seems like a mistake.

thockin replied:
@verb how bad would it be to pass a full v1.Container?

verb replied:
@thockin Not bad, It's what the kubelet does internally and I've had it working in a prototype.

We would use the API described in [Alternative 1](https://github.com/kubernetes/community/pull/649/files#diff-5cfb31b40ca47511743d0545d5697aa0R144) to POST a v1.Container (wrapped in a new top level object) to a new /debug subresource. Then the client would perform a separate /attach.

The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.

Both the API reviewers and SIG Node reviewers had concerns about using v1.Container being confusing or communicating the wrong intent to the user. Debug Containers are not general purpose containers and should not be used to build services or for routine operations. Some prefer to consider "extended exec" rather than "configurable container".

Most of the fields of v1.Container do not apply to Debug Containers and should be rejected if configured (lifecycle, livenessProbe, ports, readinessProbe, resources, stdin, stdinOnce, terminationMessagePath, terminationMessagePolicy, tty, volumeMounts). We can do this with a validation whitelist, though it would be simpler to pass a securityContext rather than a full v1.Container.

davidopp replied:
> The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.

I thought scheduler POSTs to pod/binding subresource, clients POST to pod/eviction subresource, etc. Or is there a distinction between POST and PUT here? (Which I can never keep straight.)

verb replied:
@davidopp Oh, that's good news then. When I prototyped this ~6 months ago I recall needing a couple of changes in the apiserver in order to make it work, but maybe those were specific to upgrading a connection to streaming after a POST. It's been a long ride.

Let's figure out a way to move this forward. Since we already had agreement among the reviewers at the time, and now we want to renegotiate that agreement based on new reviewers, I suggest that take the form of a new PR to amend the proposal where the new and old reviewers can work out their conflicting requirements. I'll prepare a diff.

liggitt replied:
>> The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.
>
> I thought scheduler POSTs to pod/binding subresource, clients POST to pod/eviction subresource, etc. Or is there a distinction between POST and PUT here? (Which I can never keep straight.)

Not sure about subresources, but there are issues with both PUT and POST as part of *websocket* requests (not all clients support them)

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+
+1.  `kubectl` invokes the debug API as described in the preceding section.
+1.  The API server checks for name collisions with existing containers, performs
+    admission control and proxies the connection to the kubelet's
```
note that all admission plugins that do anything related to checking containers, images, etc, would need to be updated to check ephemeral images specified in exec options now

verb replied:
This is noted in the "Security Considerations" section.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+requests and the kubelet must return an error to all but one request.
+
+There are no limits on the number of Debug Containers that can be created in a
+pod, but exceeding a pod's resource allocation may cause the pod to be evicted.
```
this also would bypass admission plugins that set resource limits/range on containers. please describe the container spec that would result from an ephemeral exec request

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+1.  `KillPod()` already operates on all running containers returned by the
+    runtime.
+1.  Containers created prior to this feature being enabled will have a
+    `containerType` of `""`. Since this does not match `"DEBUG"` the special
```
DEBUG or EPHEMERAL?

lavalamp replied:
Yes please change references in the API to "ephemeral" everywhere.

verb replied:
To be clear, this is a private label in the kubelet's runtime manager and not part of the API. I've updated it to EPHEMERAL for consistency, though.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+particular, they should enforce the same container image policy on the `Image`
+parameter as is enforced for regular containers. During the alpha phase we will
+additionally support a container image whitelist as a kubelet flag to allow
+cluster administrators to easily constraint debug container images.
```
what security context settings (uid/gid, selinux, apparmor) will the debug container have? how will admission plugins that constraint/force those (like PodSecurityPolicy) govern an ephemeral container

lavalamp replied:
I thought @verb had worked this out with @derekwaynecarr already?

verb replied:
@liggitt Compatibility with admission plugins is a top priority and a strict requirement. The implementation will depend a little bit on how the Kubernetes API settles, but it will be one of:

- The client may end up providing tje v1.Container that creates the Ephemeral Container
- If we stick with the imperative, exec-style API we can do exactly as you suggest and provide the v1.Container based on the PodExecOptions.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+*   Security Context for the Debug Container is not configurable. It will always
+    be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.
+*   Image pull policy for the Debug Container is not configurable. It will
+    always be run with `PullAlways`.
```
this prevents offline installations with pre-pulled images

verb replied:
Good point. Created https://github.com/kubernetes/kubernetes/issues/53189 to track this.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,743 @@
+        DebugStatuses []DebugStatus
+}
+
+type DebugStatus struct {
```
I don't see the necessity of this.

If you wanted to represent the way(s) in which the container is "dirty" after exec/attach/portrforward etc, I don't think this is the way to go.

verb replied:
Yes, that's something we want to represent along with details about what command was run by exec (whether traditional exec or in an ephemeral container). This isn't a blocker for the alpha implementation, though, so if you think this is the wrong approach then I'll remove this bit from the proposal and we can figure out the correct way later.

/cc @dchen1107 @thockin

thockin replied:
I think we do want some way to represent the taintedness (also should do for exec )

verb replied:
@r2d4 Can you work on the best way to represent taintedness in v1.PodStatus?

r2d4 replied:
@verb ack looking into it

ndeloof commented:
Hi @verb , I'm glad to see your repeated effort to get this feature added to k8s.

I understand the initial use-case for this effort is related to diagnostic on a running Pod, but would like to share with you another use-case which would benefit this exact same improvement:

I'm working on Jenkins integration in k8s, we run containerized builds as pods. During the build execution, a new container might be required. In many case this can be identified _before_ the build is scheduled and as such set as part of the Pod's spec, but in some cases the required image is dynamically selected as part of the build. Also, being able to run containers as part of the build just like developer do on their workstation helps to make the build script reproducible and portable.

with a new API to add (transient) containers to a Pod I could provide the glue code for build script to control such additional containers. The current usage for most users is to run a privileged (!) DinD container to host the build, or to bind mount docker.sock from host (!). As you can guess I'd prefer we don't rely on this :-\

Hope this will help understand potential use-cases this feature could support.

smarterclayton commented:
Note that I'm specifically saying:

Once ephemeral containers are available, we have the rough shape of an api
to further subdivide a pod (subcontainer division).  We could potentially
put a whole class of problems ("how do i orchestrate a set of containers
within the pod") into the same pot by having the user run a container that
spawns ephemeral containers (in whatever fashion the user wants) and using
a combination of process and container controls to manage subdivision.
I.e. instead of kube implementing a task like mechanism within a pod, make
it possible for the user to safely implement that themselves.  The
ephemeral pod API then becomes an "in pod orchestrator" which exposes the
container runtime in a kube-like fashion to the user, rather than forcing
the user to implement nested containers (which are invisible to kube) or
having kube continue to add feature after feature to approximate a directed
task graph.

On Mon, Jul 2, 2018 at 6:23 PM, Clayton Coleman <ccoleman@redhat.com> wrote:

> I've been thinking along lines like this - we don't want to add infinite
> features to Kubernetes pods that allow them to orchestrate containers, but
> we also don't want to make pods so inflexible that people build external
> container orchestration.  We have clearly stated that the pod abstraction
> is not the Borg Job/Tasks abstraction (directed graph), but allowing people
> to implement directed graph operations within a single pod has utility.
>
> If instead, a pod could leverage a node local API to add / remove
> containers within the limits the pod has defined (security boundaries,
> resources, secrets, volumes) but that the orchestration of those containers
> could be done by talking to the kubelet within the pod, then we could
> potentially continue to leverage CRI for intra-pod isolation AND avoid
> needing to add an infinite number of features to the pod api.  I think it
> deserves some thought as an approach.
>
> On Mon, Jun 25, 2018 at 3:05 AM, Nicolas De loof <notifications@github.com
> > wrote:
>
>> Hi @verb <https://github.com/verb> , I'm glad to see your repeated
>> effort to get this feature added to k8s.
>>
>> I understand the initial use-case for this effort is related to
>> diagnostic on a running Pod, but would like to share with you another
>> use-case which would benefit this exact same improvement:
>>
>> I'm working on Jenkins integration in k8s, we run containerized builds as
>> pods. During the build execution, a new container might be required. In
>> many case this can be identified *before* the build is scheduled and as
>> such set as part of the Pod's spec, but in some cases the required image is
>> dynamically selected as part of the build. Also, being able to run
>> containers as part of the build just like developer do on their workstation
>> helps to make the build script reproducible and portable.
>>
>> with a new API to add (transient) containers to a Pod I could provide the
>> glue code for build script to control such additional containers. The
>> current usage for most users is to run a privileged (!) DinD container to
>> host the build, or to bind mount docker.sock from host (!). As you can
>> guess I'd prefer we don't rely on this :-\
>>
>> Hope this will help understand potential use-cases this feature could
>> support.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/community/pull/649#issuecomment-399852982>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_pwgpJIR1A7P9-1qdHkSNueZ5QXiQks5uAIvDgaJpZM4NinaN>
>> .
>>
>
>

avanier commented:
I have a keen interest for this feature. Does anyone have movement to report on this outside of this ticket?

Also, to echo @ndeloof mentions for debuging Jenkins, I have to say this is indeed quite useful. I've been running [Concourse](https://concourse-ci.org/) at a few shops over recent years, and using Pivotal's Garden OCI runtime they achieve exactly that.

verb commented:
@avanier kubernetes/features#277 might be better for tracking progress of this feature. The API change is under review in kubernetes/kubernetes#59416, once the API changes there should be quick progress.
